### PR TITLE
[NO MERGE] Ray postprocessing and writes. kinda goes zoom.

### PR DESCRIPTION
**Pull request by jacobsela**
fully offload post processing and writes.

Loading and offloading tensors from GPU is still on main process. Not 100% how (and if we should) get them off. At this point it may be better to just scale horizontally with DOs.

```python
from time import time_ns

import fiftyone as fo
import fiftyone.zoo as foz

dataset = fo.load_dataset("coco-val")
classes = dataset.distinct("ground_truth.detections.label")
model = foz.load_zoo_model("omdet-turbo-swin-tiny-torch", classes=classes)

start = time_ns()
dataset.apply_model(
    model,
    label_field="predictions_ray",
    batch_size=16,
    num_workers=32,
    confidence_thresh=0.1,
    skip_failures=False,
)
end = time_ns()

print(f"Apply Model Time: {(end - start) / 1e9} seconds") # Apply Model Time: 147.059405114 seconds
```

Screenshot 2025-08-25 at 7 10 53 PM

all remaining time in `predict_all` that isn't forward pass is moving about tensors from cpu to gpu or back.
Screenshot 2025-08-25 at 7 20 01 PM

I guess if anything this is where multithreading would shine. Not 100% sure how to integrate it here though...

## Summary by CodeRabbit

- New Features
  - Parallelized image model inference and label writing for faster processing on large datasets
  - Automatic initialization of the distributed runtime when using parallel processing
- Refactor
  - Improved propagation of sample IDs through data pipelines to correctly associate predictions
  - Enhanced handling of model outputs without an output processor, including CPU-safe tensors and image size returns
  - More flexible batching in transformer pipelines, allowing non-tensor fields to pass through without padding

### CODE REVIEW

File: fiftyone/core/ray/base.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
+import ray
+
+import fiftyone.core.view as fov
+
+
+def serialize_samples(samples):
```
Comment (kaixi-wang) [inline fiftyone/core/ray/base.py:6]: This is not a good pattern because you're guaranteeing that the process will require a database connection just to resolve the file path. It would be much better to just resolve the file path directly

File: fiftyone/core/ray/writers.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+        if self.post_processor is not None:
+            payloads = self.post_processor(
+                *payloads, confidence_thresh=self.confidence_thresh
+            )
+
+        with self.ctx:
+            for sample, payload in zip(samples_batch, payloads):
+                sample.add_labels(
+                    payload,
```
Comment (kaixi-wang) [inline fiftyone/core/ray/writers.py:36]: This is what actually needs to be refactored.

**Review by kaixi-wang — CHANGES_REQUESTED** [review]
This adds Ray as a dependency and introduces the complexity of relying on another opinionated library but doesn't solve the root of the issue.

**Comment by exupero** [comment]
Not being familiar with Ray, I have a couple high-level questions:
1. Is this using a background thread or process to offload work? (Or does Ray decide?)
2. If using a background process, does this get around the problem of SampleView and SaveContext not being picklable?

File: fiftyone/core/ray/base.py (hunk 1 above)
Comment (jacobsela) [reply fiftyone/core/ray/base.py:6]: What file paths are we talking about?

Generally speaking, we want these workers to have a database connection. One of the goals is for them to interact with FO datasets in parallel to the main process.

File: fiftyone/core/ray/writers.py (hunk 2 above)
Comment (jacobsela) [reply fiftyone/core/ray/writers.py:36]: Any optimizations to the base code are welcome. That said, notice that we are offloading multiple things here from the main process, namely:
1. model output post processing
2. `add_labels`
3. writing to mongo (with `ctx.save()`)

all of these parts have some impact on final `apply_model` time.

**Comment by jacobsela** [comment]
@kaixi-wang 

Re:

> This adds Ray as a dependency and introduces the complexity of relying on another opinionated library but doesn't solve the root of the issue.

There are two parts here:
*(1) the underlying issue*
The core problem this PR tries to address is offloading cpu-bound and I/O-bound operations from the main process, which we want to be only gpu-bound. It seems to do so, as indicated by the profiling attached. I am assuming that by underlying issue you mean the inherent performance of things like `add_labels`. Even if `add_labels` was optimized to take 0 time, there are other parts of the code that are a GPU utilization bottlenecks. An approach like this allows us to more easily offload _whatever_ we view as the bottleneck from any process. More generally, local optimizations can be really impactful for frequently called, especially slow operations. However, if we are interested in longer term scaling, our main concern should be dividing the code to pieces that we can individually scale according to a workload's needs.

This leads well to the second point, the use of ray (which is built to do exactly this).

*(2) adding ray as a dependency*
Yes, this is a decision we will have to make. Ray is just a means to an end here. Multiprocessing is a useful tool that can be helpful throughout our codebase. By introducing something like Ray, I hope to cut out a lot of development time that we would otherwise spend on making multiprocessing use in FiftyOne consistent, trusted, clean, and efficient. What are strong arguments for *not* using ray from your point of view? Let's discuss them and come to a decision.

File: fiftyone/utils/torch.py [hunk 3]
```diff
@@ -921,9 +922,9 @@ def _predict_all(self, imgs):
 
         if self._output_processor is None:
             if isinstance(output, torch.Tensor):
-                output = output.detach().cpu().numpy()
+                output = output.detach().cpu()
```
Comment (jacobsela) [inline fiftyone/utils/torch.py:925]: notice this

File: fiftyone/core/ray/base.py (hunk 1 above)
Comment (kaixi-wang) [reply fiftyone/core/ray/base.py:6]: No you don't want to give workers database connections. There is zero benefit because the data (media) is not even in the database and cannot be retrieved using the database connection

File: fiftyone/utils/transformers.py [hunk 4]
```diff
@@ -681,7 +681,17 @@ def _predict_all(self, args):
             )
 
         else:
-            return output
+            for k, v in output.items():
```
Comment (jacobsela) [inline fiftyone/utils/transformers.py:684]: this

File: fiftyone/core/ray/base.py (hunk 1 above)
Comment (jacobsela) [reply fiftyone/core/ray/base.py:6]: These are writer workers. They must each hold some connection or share access to a pool of connections for us to write multiple things in parallel. Frankly we don't even need a ton of them because most of what they do is sit around waiting for I/O, so a single one that's multithreaded would probably be just fine.

Unrelated grievances with our multi-worker read system can be discussed elsewhere.
